### ISSUE TITLE
Figure out how to route logs in the Azure Logs integration

### ISSUE THREAD

**Issue by zmoog** [zmoog/public-notes#92]
## Situation

The Azure Logs integration allows multiple log categories to be collected from a single event hub.

At a high level, users (1) define the event hub name and settings, and (2) the integration will use the same event hub for all the integrations.

![CleanShot 2024-08-28 at 23 10 05@2x](https://github.com/user-attachments/assets/f5f1b040-20be-4a01-9e8e-0ecffb76cf52)

## Problem

This setup is inefficient, and we plan to change it in future releases.

## Solutions

You can only use the generic integration and route logs to the right data stream using the reroute processor.

![image](https://github.com/user-attachments/assets/e9a7fa38-18a4-4eb2-a590-46de135cdd5b)

**Comment by zmoog** [comment]
With one input + routing, we can reduce the user errors metric to zero and make the fewest storage account API calls possible.

Here's the diagram to leverage routing:

![image](https://github.com/user-attachments/assets/5d61cdce-bf02-4d5e-93ac-e11132057be1)


1. One input receives log events from the event hub
2. The input publishes the log events to the `logs-azure.eventhub-default` data stream.
3. The `logs-azure.eventhub-default` data stream contains a `logs-azure.eventhub@custom` custom pipeline with rules to route log events based on the log category.
4. Each log event lands in the target data stream.

If the routing rules cover all incoming log categories, the `logs-azure.eventhub-default` data stream will be empty. However, we can set up an alarm rule to trigger a notification in case any log event doesn't have a routing rule so that we can iterate and update the `logs-azure.eventhub@custom` custom pipeline. 

The routing option is probably the most efficient method.

Here's the source code of the `logs-azure.eventhub@custom` pipeline I am testing:

```
PUT _ingest/pipeline/logs-azure.eventhub@custom
{
  "processors": [
    {
      "json": {
        "field": "message",
        "target_field": "tmp_json"
      }
    },
    {
      "set": {
        "field": "routing_category",
        "copy_from": "tmp_json.category",
        "ignore_empty_value": true
      }
    },
    {
      "remove": {
        "field": "tmp_json",
        "ignore_missing": true
      }
    },
    {
      "reroute": {
        "dataset": [
          "azure.signinlogs"
        ],
        "if": "ctx.routing_category == \"SignInLogs\" || ctx.routing_category == \"NonInteractiveUserSignInLogs\" || ctx.routing_category == \"ServicePrincipalSignInLogs\" || ctx.routing_category == \"ManagedIdentitySignInLogs\""
      }
    },
    {
      "reroute": {
        "dataset": [
          "azure.identity_protection"
        ],
        "if": "ctx.routing_category == \"RiskyUsers\" || ctx.routing_category == \"UserRiskEvents\""
      }
    },
    {
      "reroute": {
        "dataset": [
          "azure.provisioning"
        ],
        "if": "ctx.routing_category == \"ProvisioningLogs\""
      }
    },
    {
      "reroute": {
        "dataset": [
          "azure.auditlogs"
        ],
        "if": "ctx.routing_category == \"AuditLogs\""
      }
    },
    {
      "reroute": {
        "dataset": [
          "azure.activitylogs"
        ],
        "if": "ctx.routing_category == \"Administrative\" || ctx.routing_category == \"Security\" || ctx.routing_category == \"ServiceHealth\" || ctx.routing_category == \"Alert\" || ctx.routing_category == \"Recommendation\" || ctx.routing_category == \"Policy\" || ctx.routing_category == \"Autoscale\" || ctx.routing_category == \"ResourceHealth\""
      }
    },
    {
      "reroute": {
        "dataset": [
          "azure.graphactivitylogs"
        ],
        "if": "ctx.routing_category == \"MicrosoftGraphActivityLogs\""
      }
    },
    {
      "reroute": {
        "dataset": [
          "azure.firewall_logs"
        ],
        "if": "ctx.routing_category == \"AzureFirewallApplicationRule\" || ctx.routing_category == \"AzureFirewallNetworkRule\" || ctx.routing_category == \"AzureFirewallDnsProxy\" || ctx.routing_category == \"AZFWApplicationRule\" || ctx.routing_category == \"AZFWNetworkRule\" || ctx.routing_category == \"AZFWNatRule\" || ctx.routing_category == \"AZFWDnsQuery\""
      }
    },
    {
      "reroute": {
        "dataset": [
          "azure.application_gateway"
        ],
        "if": "ctx.routing_category == \"ApplicationGatewayFirewallLog\" || ctx.routing_category == \"ApplicationGatewayAccessLog\""
      }
    }
  ]
}
```

**Comment by nicpenning** [comment]
That seems clean to me.

FYI :  Your second diagram is showing as missing.
![image](https://github.com/user-attachments/assets/74c7dd7b-dfe3-4372-b827-ce11bf2ae764)

**Comment by zmoog** [comment]
> FYI : Your second diagram is showing as missing.

Ouch, I probably copied and pasted an expiring URL from GitHub. Checking!

**Comment by zmoog** [comment]
It should be fixed now.

**Comment by nicpenning** [comment]
How does this model work if you wanted more than 1 agent for redundancy and improved performance?

**Comment by zmoog** [comment]
> How does this model work if you wanted more than 1 agent for redundancy and improved performance?

Good question! I should update the note to add this detail.

Here is a diagram showing how the two inputs work together to achieve improved redundancy and performance.

![CleanShot 2024-09-09 at 10 57 08](https://github.com/user-attachments/assets/5a0a9981-f81e-45b0-8878-17a9da6ab550)

Users set up diagnostic settings, sending data to an event hub (1). The two (or more) inputs start and claim an equal part of partitions. With a four-partition event hub, two inputs usually get two partitions each. Each input processes messages and sends them to the data stream in Elasticsearch.

The routing (2) happens on Elasticsearch at the data stream level, so it works with one or multiple event hubs.

**Comment by nicpenning** [comment]
This sounds great. Unfortunately the graphic won't load for me.

**Comment by nicpenning** [comment]
I can zoom in here, looks awesome!

**Comment by zmoog** [comment]
> I can zoom in here, looks awesome!

Yeah, the GitHub images URL expires quickly. I usually reload the page and click on the image to get the whole picture. Let me know if you have difficulties in opening it.

**Comment by nicpenning** [comment]
Works great now.

**Comment by threatangler-jp** [comment]
Also, important to note here this issue with the azure integration caused a 2x increase in the eventhub expenses in azure. The eventhub costs increased by 2x in each azure environment when we upgraded and the memory leak started. Then those costs dropped as soon as we resolved the memory leak.

**Comment by zmoog** [comment]
> Also, important to note here this issue with the azure integration caused a 2x increase in the eventhub expenses in azure. The eventhub costs increased by 2x in each azure environment when we upgraded and the memory leak started. Then those costs dropped as soon as we resolved the memory leak.

Yeah, this appeared after the same library upgrade that brought in the memory leak.

I am self-quoting with same comment from the [Elastic Community Slack](https://www.elastic.co/blog/join-our-elastic-stack-workspace-on-slack):

> The root cause is similar to the memory leak. The contention [among consumers in v1 integrations] causes the disconnections, which causes the input to do more work reading and storing the partition checkpoint information, making more API calls to the Storage Account.
> 
> However, the cost increase is not proportional to the additional reads/writes. There is probably another bug in the MS libraries that causes the number of API calls to increase linearly over time.
> 
> For example, suppose the contention among consumers causes 10 new storage account API calls every minute.
> 
> In the first minute, the input makes 10 API calls. You expect that in the second minute, it would make 10 additional API calls.
> 
> Instead, the input makes:
> 
> - minute 1: 10
> - minute 2: 15
> - minute 3: 20
> - and so on.
> 
> Removing the contention and disconnects means all these edge cases do not manifest, and the input can run for an extended period with no issues.

I recommend switching to the Azure Logs integration v2 preview that completely removes the consumers contention.
